In [66]:
import math
"""def calculate_density(C, time, step_size, starting_value):
    for i in range(1, int(time/step_size)+1):
        starting_value += C*starting_value*step_size
#        print(starting_value)
    return starting_value
print(calculate_density(1.03526, 1, 0.1, 10))
exp_growth = lambda time, double, C: math.exp(math.log(time/double)*C)
#print(exp_growth(10, 10,2))"""
start = 10
double_time = 20
time = 1
C = (math.log(2)/double_time)
print(C)
x_final = 80
t_for_rep = math.log(x_final / start) / C
print(t_for_rep)
x_final = 100
t_for_rep = math.log(x_final / start) / C
print(t_for_rep)

0.034657359028
60.0
66.4385618977
100.0


Integrating $\frac{\delta x}{\delta t} = C x$ gives $x_{t} = x_0 \cdot e^{C t}$.
We know that $x_{0} = 10$ and $x_{20} = 20$, therefore: $20 = 10 \cdot e^{C \cdot 20} \Leftrightarrow 2 = e^{C \cdot 20} \Leftrightarrow log(2) = C \cdot 20 \Leftrightarrow C = \frac{ln(2)}{20} \Leftrightarrow C = 0.03466$.

Or more generally $C = \frac{log(2)}{D}$ where D is the doubling time.

To calculate when a certain increase from the original number has occured we can will in a corresponding multiple of the starting value and solve the formula for $t$: $x_{final} = x_0 \cdot e^{C \cdot t} \Leftrightarrow \frac{x_{final}}{x_0} = e^{C \cdot t} \Leftrightarrow ln(\frac{x_{final}}{x_0}) = C \cdot t \Leftrightarrow t = \frac{ln(\frac{x_{final}}{x_0})}{C}$
In the case of $x_{final} = 80$ this gives $t = 60$ and $x_{final} = 100$ results in $t = 66.4386$.

In [24]:
%matplotlib inline
from ipywidgets import interact, interactive
from IPython.display import clear_output, display, HTML

#import numpy as np
#from scipy import integrate
import math
from matplotlib import pyplot as plt
#from mpl_toolkits.mplot3d import Axes3D
#from matplotlib.colors import cnames
#from matplotlib import animation


def SIR(mu, beta_0, season_length, gamma, I_0, time):
    delta_t = 0.1
    S = 1-I_0
    I = I_0
    R = 0.0
    S_list = []
    I_list = []
    R_list = []
    time_list = []
    for t in range(int(time/delta_t)):
        S_list.append(S)
        I_list.append(I)
        R_list.append(R)
        time_list.append(t*delta_t)
        S_new = S + (mu - mu * S - beta_0 * (1+ math.sin(2*math.pi*t*delta_t/season_length))*I*S) * delta_t
        I_new = I + (beta_0 * (1+ math.sin(2*math.pi*t*delta_t/season_length))*I*S - (gamma+mu) * I) * delta_t
        R_new = R + (gamma * I - mu * R) * delta_t
        S, I, R = S_new, I_new, R_new
    plt.figure()
    plt.plot(time_list, S_list, "r", label = "S")
    plt.plot(time_list, I_list, "b", label = "I")
    plt.plot(time_list, R_list, "g", label = "R")
    plt.xlabel("Time")
    plt.ylabel("Fraction of population")
    plt.legend(loc = "upper right")
    plt.show()
SIR_interact = interactive(SIR, mu=(0.0, 1.0), beta_0=(0.0, 1.0), 
                           season_length=(0.0, 30.0), gamma=(0.0, 1.0), I_0=(0.0, 1.0), time =(10,110))
display(SIR_interact)


aW50ZXJhY3RpdmUoY2hpbGRyZW49KEZsb2F0U2xpZGVyKHZhbHVlPTAuNSwgZGVzY3JpcHRpb249dSdtdScsIG1heD0xLjApLCBGbG9hdFNsaWRlcih2YWx1ZT0wLjUsIGRlc2NyaXB0aW9uPXXigKY=
